### ISSUE TITLE
`redundant guard` on match guard of an expression rebinding

### ISSUE THREAD

**Issue by Thaumy** [rust-lang/rust-clippy#11841]
### Summary

When matching an expression with a guard that is true only if the expression's rebinding does not have a certain value, clippy will warn the guard as `redundant guard` and its help message will change the code's original meaning.

### Reproducer

I tried this code:

```rust
fn main() {
    match 1 + 1 {
        b if match b {
            1 => false,
            _ => true
        } => println!("branch 1"),
        _ => println!("branch 2")
    }
}
```

I expected to see this happen:

No warning of `redundant guard`

Instead, this happened:

```text
warning: redundant guard
 --> src/main.rs:3:14
  |
3 |           b if match b {
  |  ______________^
4 | |             1 => false,
5 | |             _ => true
6 | |         } => println!("branch 1"),
  | |_________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_guards
  = note: `#[warn(clippy::redundant_guards)]` on by default
help: try
  |
3 -         b if match b {
3 +         1 => println!("branch 1"),
  |
```

### Version

```text
rustc 1.76.0-nightly (283170175 2023-11-17)
binary: rustc
commit-hash: 2831701757eb7b3105eda26a306c2f3a97e2664b
commit-date: 2023-11-17
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.5
```

**Comment by Centri3** [comment]
This looks suspiciously like `matches!`, but flipped. We should check that the wildcard arm is `false` and the "matches" arm is `true`, and not lint if so
